### DISCUSSION TITLE
Build using llvm-config instead of $prefix/lib/cmake/llvm

### DISCUSSION THREAD

**Discussion by jedbrown** [EnzymeAD/Enzyme#1667] [General]
When the path in `-DLLVM_DIR=/path/to/llvm` does not exist or does not contain the necessary CMake files, Enzyme silently falls back to a system version (if it exists). This is the case when building for Rust, in which `$prefix/lib` contains only `libLLVM-17-rust-1.77.0-nightly.so` (no `cmake/llvm` subdirectory). It would be a bit of a chore to ensure these CMake files are available in all contexts because [sysroot is distributed multiple ways](https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html#what-is-a-sysroot).

> `libLLVM.so`, the shared object file for the LLVM project. In-tree this is either built from source or downloaded from CI; out-of-tree, you need to install llvm-tools-preview with rustup.

However, `llvm-config` is always available. How hard would it be for Enzyme's build to configure itself entirely from `llvm-config`, without the CMake files for LLVM?

**Comment by wsmoses** [comment]
I'm okay with that, though I'm not sure how difficult it would be a-priori.

I'm sure someone has done this in the past for LLVM, so perhaps we can find something?

**Reply by jedbrown** [reply]
CHIP-SPV has [a start](https://github.com/CHIP-SPV/chipStar/blob/main/cmake/FindLLVM.cmake), though it looks like it only needs command-line tools. I see Enzyme's CMakeLists.txt has `include(AddLLVM)`, which is a big file. If we only needed the variables like `LLVM_DEFINITIONS`, `LLVM_INCLUDE_DIRS`, etc., then it would be pretty easy. There is `src/llvm-project/llvm/cmake/modules/AddLLVM.cmake` inside the rust repo so perhaps it can be called, but I've never seen this sort of CMake task not snowball.

**Comment by vchuravy** [comment]
> However, `llvm-config` is always available

That is not the case for cross-compilation `llvm-config` is only available on the host and my be wrong for the target system.

**Reply by jedbrown** [reply]
Interesting, I'm trying to work out whether that is important for Rust. During bootstrap, stage 1 is always for the host. I don't think it's documented how a target libLLVM is produced when cross-compiling, but we must eventually need it.

I don't think we would ever need to execute `llvm-config` on the target. When executed on the host, is `llvm-config` unable to specify how to build for the target?

**Reply by vchuravy** [reply]
> When executed on the host, is llvm-config unable to specify how to build for the target?

Sadly. `llvm-config` only knows about the llvm build it comes from, not the one you are creating. There might be ways around this with a staged build, but those have been very fragile for us and for binarybuilder/Yggdrasil we went with relying on the cmake files.
